rfcs/2020-11-02-remap-infallible-functions.md

```diff
@@ -0,0 +1,371 @@
+provide more runtime guarantees to users of the Remap language.
+
+## Scope
+
+- The RFC describes how users can call both fallible and infallible variants of
+  functions.
+- It does so through a more broader context of fallible and infallible
+  _expressions_
+- It _does not_ decide on which expressions (other than functions) should
+  support the concept of infallibility.
+
+## Motivation
+
+As we expand the use-case for the Remap language within Vector, we want to be
+able to guide users as much as possible as they have their first experience with
+the language in Vector.
+
+One way to do so is to have programs run _as predictable as possible_. This
+means moving errors as much as possible to boot-time, while reducing the list of
+potential runtime exceptions.
+
+This RFC proposes a way to allow users to determine how they want to handle
+runtime errors for each individual function, so that they can decide if an error
+is truly an exception and should abort program execution, or if the error can
+safely be ignored.
+
+## Internal Proposal
+
+The following Remap program is easy to understand:
+
+```javascript
+.message = parse_json(.message)
+.timestamp = now()
+```
+
+It:
+
+1. Reads the value of the `message` field of the object.
+2. Parses the value to a JSON value (a literal, array, or map).
+3. Assigns the new value to the `message` field, removing the old value.
+4. Assigns the current timestamp to the `timestamp` field.
+
+### Fallible Functions
+
+While the program itself is easy to understand, running it can result in
+unexpected situations.
+
+If whatever value the `message` field contains cannot be parsed as valid JSON,
+the `parse_json` function returns an error.
+
+This results in an outcome where the original event is left untouched, the
+`message` field still contains the original value, and no `timestamp` field
+exists (assuming none existed to start with).
+
+A use-case such as the above is quite common. Fields often need to be parsed as
+JSON values, but at the same time those fields are highly dynamic and might not
+always be considered valid JSON, for example if a bug in an application starts
+generating invalid stringified JSON output.
+
+Users need a way to handle such situations, without having to resort to complex
+solutions.
+
+### Infallible Function Variants
+
+The solution proposed by this RFC is to introduce the concept of _infallible
+function variants_.
+
+An infallible function variant is an exact copy of an existing fallible
+function, but it is guaranteed to never fails.
+
+The infallible function must be as easy (or easier) to call than the fallible
+function, without any significant syntactic complexity.
+
+### To Bang Or Not To Bang
+
+The proposal is to introduce the concept of the bang (`!`) function identifier
+additive to determine as a user if a function should be allowed to fail or not.
+
+Function identifiers with a bang are considered to be fallible, while functions
+without are not:
```

Comment:
As someone with a ruby background, I find this a bit confusing. There is a convention in that community that functions named with a bang will raise an exception on failure, while non-bang functions return nil. This makes sense to me because the `!` bring attention to the fact that it may blow up. It also makes sense in the case of remap because it feels like a context in which we should be forgiving of imperfect data.

If we do go with the bang idea (I'm not yet convinced), then I'd recommend reversing the meaning.

Reply:
Did someone say Ruby? 😄